extension/src/openvic-extension/singletons/CursorSingleton.cpp

```diff
@@ -0,0 +1,622 @@
+	
+	
+
+	//there is also a png file in the folder we don't want to bother loading
+	//so don't just load everything in the directory
+
+	//We need to load both ".cur" and ".CUR" files
+	Dataloader::path_vector_t cursor_files = game_singleton->get_dataloader()
+		.lookup_files_in_dir_recursive(cursor_directory, ".cur");
+
+	Dataloader::path_vector_t CURsor_files = game_singleton->get_dataloader()
+		.lookup_files_in_dir_recursive(cursor_directory, ".CUR");
+	cursor_files.insert(std::end(cursor_files),std::begin(CURsor_files),std::end(CURsor_files));
+
+	Dataloader::path_vector_t animated_cursor_files = game_singleton->get_dataloader()
+		.lookup_files_in_dir_recursive(cursor_directory, ".ani");
+	
+	if (cursor_files.empty() && animated_cursor_files.empty()){
+		Logger::error("failed to load cursors: no files in cursors directory");
+		return false;
+	}
+
+	bool ret = true;
+
+	for(fs::path const& file_name : cursor_files) {
+		String file = Utilities::std_to_godot_string(file_name.string());
+		StringName name = to_define_file_name(file);
+
+		if (!_load_cursor_cur(name,file)){
+			Logger::error("failed to load normal cursor at path ", file_name);
+			ret = false;
+		}
+	}
+
+	for(fs::path const& file_name : animated_cursor_files) {
+		String file = Utilities::std_to_godot_string(file_name.string());
+		StringName name = to_define_file_name(file);
+
+		if (!_load_cursor_ani(name,file)){
+			Logger::error("failed to load animated cursor at path ", file_name);
+			ret = false;
+		}
+	}
+
+	return ret;
+}
+
+static int32_t _reverse_bits(int32_t byte, int32_t bitsPerPixel=8) {
+	int32_t reverser_lookup[] = {
+		0x0, 0x8, 0x4, 0xc, 0x2, 0xa, 0x6, 0xe,
+		0x1, 0x9, 0x5, 0xd, 0x3, 0xb, 0x7, 0xf
+	};
+	int32_t a = reverser_lookup[(byte & 0b1111)] << 4;
+	int32_t b = reverser_lookup[byte >> 4];
+	int32_t c = b | a;
+	return c >> (8-bitsPerPixel);
+}
+
+static int32_t _rotate_right(int32_t byte, int32_t size=8) {
+	return ((byte & 0b1) << (size-1)) | (byte >> 1);
+}
+
+static int32_t _load_int_256(Ref<FileAccess> const& file) {
+	int32_t value = file->get_8();
+	if (value == 0) value = 256;
+	return value;
+}
+
+static int32_t _get_row_start(Vector2i dimensions, int32_t y_coord, int32_t bitsPerPixel) {
+	int32_t rowCount = dimensions.x * bitsPerPixel / 32;
+	bool hasExtraRow = ((dimensions.x * bitsPerPixel) % 32) != 0;
+	rowCount += hasExtraRow; //TODO 1*
+	return rowCount*y_coord*4; //4 bytes per row * rows down
+}
+
+static int32_t _select_bits(PackedByteArray const& data, int32_t rowStart, int32_t firstBit, int32_t bitCount) {
+	int32_t bitmasks[] = {0b1 , 0b11, 0b111, 0b1111, 0b11111, 0b111111, 0b1111111, 0b11111111};
+	int32_t byteIndex = firstBit / 8;
+	int32_t bitInByteIndex = firstBit % 8;
+	if (bitInByteIndex + bitCount > 8){
+		Logger::error("Attempted to select bits outside of a byte.");
+		return 0;
+	}
+	int32_t byte = _reverse_bits(data[rowStart + byteIndex]);
+	int32_t selected = (byte >> bitInByteIndex) & bitmasks[bitCount-1];
```

Comment:
```suggestion
static constexpr int32_t _select_bits(PackedByteArray const& data, int32_t rowStart, int32_t firstBit, int32_t bitCount) {
	int32_t byteIndex = firstBit >> 3;
	int32_t bitInByteIndex = firstBit & 0b111;
	if (bitInByteIndex + bitCount > 8) {
		Logger::error("Attempted to select bits outside of a byte.");
		return 0;
	}
	int32_t byte = _reverse_bits(data[rowStart + byteIndex]);
	int32_t selected = (byte >> bitInByteIndex) & ((1 << bitCount) - 1);
```
- If we were to stick with the `bitmasks` array then it should be `static constexpr`, but its values are equivalent to the equation above so we might as well just use it directly and avoid the array in the first place.
- We can replace `/ 8` and `% 8` with bitwise equivalents: `>> 3` and `& 0b111`.
- Also it seems we can actually make this `constexpr` as it only uses the `PackedByteArray` as a `const&` without actually constructing or destructing it or using any of its other functions, although I'm not sure why using its `[]` operator isn't causing any issues (maybe it's treated as a function call to a non-constexpr part and the rest of the function around it is `constexpr`, similar to how `Logger` calls are allowed?)

I believe you could make even more functions `constexpr` by replacing uses of `Vector2i` in arguments with separate `int32_t x, int32_t y` arguments (or just one if you don't need both), and maybe also replacing `PackedByteArray const&` with `uint8_t const*`, got using `PackedByteArray`'s `ptr()` function (I do wonder if we should be testing against the array's length, and similarly checking the length of arrays got from files to make sure they're as long as we expect. It's fine if the code is guarunteed not to look outside of the array's bounds, but if there's a possibility, even just from incorrectly formatted files, then we should add some checks to avoid potential crashes)

Reply:
The [] operator does seem to make the build fail on linux.